include/merlin/core_kernels.cuh

```diff
@@ -169,11 +163,19 @@ void initialize_buckets(Table<K, V, S>** table, const size_t start,
   }
 
   (*table)->num_of_memory_slices += num_of_memory_slices;
+  uint32_t bucket_max_size = static_cast<uint32_t>((*table)->bucket_max_size);
+  size_t bucket_memory_size =
+      bucket_max_size * (sizeof(AtomicKey<K>) + sizeof(AtomicScore<S>));
+  // Align the digests according to the cache line size.
+  uint32_t reserve_digest =
+      ((bucket_max_size + 127) / 128) * 128 * sizeof(uint8_t);
+  bucket_memory_size += reserve_digest;
   for (int i = start; i < end; i++) {
-    CUDA_CHECK(cudaMalloc(&((*table)->buckets[i].keys_),
-                          (*table)->bucket_max_size * sizeof(AtomicKey<K>)));
-    CUDA_CHECK(cudaMalloc(&((*table)->buckets[i].scores_),
-                          (*table)->bucket_max_size * sizeof(AtomicScore<S>)));
+    CUDA_CHECK(cudaMalloc(&((*table)->buckets[i].keys_), bucket_memory_size));
+    (*table)->buckets[i].scores_ = reinterpret_cast<AtomicScore<S>*>(
+        (*table)->buckets[i].keys_ + bucket_max_size);
+    (*table)->buckets[i].digests = reinterpret_cast<uint8_t*>(
```

Comment:
If we need to bring the `digests` to be ahead of `keys_`, the `find` should always read the `digests` first.

Reply:
OK, its reasonable

Reply:
And I think we can just store the key and 's addresses, as the addresses of scores and digests can be infered by keys_ andbucket_max_size